checks.d/snmp.py

```diff
@@ -200,53 +202,60 @@ def check_table(self, instance, oids, lookup_names, timeout, retries):
-                    missing_results.append(oid)
-                else:
-                    complete_results.append(var)
-
-            if missing_results:
-                # If we didn't catch the metric using snmpget, try snmpnext
-                error_indication, error_status, error_index, var_binds_table = self.snmpgetnext(
+            try:
+                # Start with snmpget command
+                error_indication, error_status, error_index, var_binds = self.snmpget(
                     auth_data,
                     transport_target,
-                    *missing_results,
-                    lookupValues=lookup_names,
+                    *(oids[first_oid:first_oid + self.oid_batch_size]),
+                    lookupValues=enforce_constraints,
                     lookupNames=lookup_names)
 
                 # Raise on error_indication
                 self.raise_on_error_indication(error_indication, instance)
 
-                if error_status:
-                    message = "{0} for instance {1}".format(error_status.prettyPrint(),
-                                                            instance["ip_address"])
-                    instance["service_check_error"] = message
-                    self.warning(message)
+                missing_results = []
+                complete_results = []
 
-                for table_row in var_binds_table:
-                    complete_results.extend(table_row)
+                for var in var_binds:
+                    result_oid, value = var
+                    if reply_invalid(value):
+                        oid_tuple = result_oid.asTuple()
+                        oid = ".".join([str(i) for i in oid_tuple])
+                        missing_results.append(oid)
+                    else:
+                        complete_results.append(var)
+
+                if missing_results:
+                    # If we didn't catch the metric using snmpget, try snmpnext
+                    error_indication, error_status, error_index, var_binds_table = self.snmpgetnext(
+                        auth_data,
+                        transport_target,
+                        *missing_results,
+                        lookupValues=enforce_constraints,
+                        lookupNames=lookup_names)
+
+                    # Raise on error_indication
+                    self.raise_on_error_indication(error_indication, instance)
+
+                    if error_status:
+                        message = "{0} for instance {1}".format(error_status.prettyPrint(),
+                                                                instance["ip_address"])
+                        instance["service_check_error"] = message
+                        self.warning(message)
+
+                    for table_row in var_binds_table:
+                        complete_results.extend(table_row)
+
+                all_binds.extend(complete_results)
+            except PySnmpError as e:
```

Comment:
This could happen if we *do* actually enforce constraints and a particular OID does not match it, like an agent returning 0 for a value that should be 1..65535 - then that batch would fail.... but we could probably still try to collect metrics from the other batches.